gaboon/commands/install.py

```diff
@@ -1,49 +1,301 @@
+    progress_bar = tqdm(total=total_size, unit="iB", unit_scale=True)
+    content = bytes()
+
+    for data in response.iter_content(1024, decode_unicode=True):
+        progress_bar.update(len(data))
+        content += data
+    progress_bar.close()
+
+    with zipfile.ZipFile(BytesIO(content)) as zf:
+        zf.extractall(target_path)
+
+
+def _maybe_retrieve_github_auth() -> dict[str, str]:
+    """Returns appropriate github authorization headers.
+
+    Otherwise returns an empty dict if no auth token is present.
+    """
+    token = os.getenv("GITHUB_TOKEN")
+    if token:
+        auth = b64encode(token.encode()).decode()
+        return {"Authorization": f"Basic {auth}"}
+    return {}
+
+
+def _get_download_url_from_tag(org: str, repo: str, version: str, headers: dict) -> str:
+    response = requests.get(
+        f"https://api.github.com/repos/{org}/{repo}/tags?per_page=100", headers=headers
+    )
+    if response.status_code != 200:
+        msg = "Status {} when getting package versions from Github: '{}'".format(
+            response.status_code, response.json()["message"]
+        )
+        if response.status_code in (403, 404):
+            msg += (
+                "\n\nMissing or forbidden.\n"
+                "If this issue persists, generate a Github API token and store"
+                " it as the environment variable `GITHUB_TOKEN`:\n"
+                "https://github.blog/2013-05-16-personal-api-tokens/"
             )
+        raise ConnectionError(msg)
+
+    data = response.json()
+    if not data:
+        raise ValueError("Github repository has no tags set")
+    org, repo = data[0]["zipball_url"].split("/")[3:5]
+    tags = [i["name"].lstrip("v") for i in data]
+    if version not in tags:
+        raise ValueError(
+            "Invalid version for this package. Available versions are:\n"
+            + ", ".join(tags)
+        ) from None
+    return next(i["zipball_url"] for i in data if i["name"].lstrip("v") == version)
 
+
+def _pip_installs(
+    package_ids: list[str], base_install_path: Path, installer: str, quiet: bool = False
+):
+    logger.info(f"Installing {len(package_ids)} pip packages...")
+    cmd = []
     # TODO: Allow for multiple versions of the same package to be installed
-    cmd = ["uv", "pip", "install", *package_ids, "--target", str(path)]
+    if installer == "uv":
+        cmd = ["uv", "pip", "install", *package_ids, "--target", str(base_install_path)]
+    else:
+        cmd = ["pip", "install", *package_ids, "--target", str(base_install_path)]
 
     # TODO: `--upgrade` and `--force` options.
     capture_output = quiet
-    subprocess.run(cmd, capture_output=capture_output, check=True)
+    try:
+        subprocess.run(cmd, capture_output=capture_output, check=True)
+    except FileNotFoundError as e:
+        logger.info(
+            f"Stack trace:\n{''.join(traceback.format_exception(type(e), e, e.__traceback__))}"
+        )
+        logger.error(e)
```

Comment:
these basically print the same thing right?

Reply:
No, `logger.error(e)` just prints the error, like:

```
ValueError: blah blah blah
```

The other line prints the actual stack trace:
```
/path/thing/
- line
/path/thing/
- line
```